### ISSUE TITLE
Decide what to do about '/'

### ISSUE THREAD

**Issue by GoogleCodeExporter** [google/gson#356]
```
One of our users correctly pointed out that we're disagreeing with the standard 
in our treatment of the '/' character. 

https://groups.google.com/forum/#!topic/google-gson/c77R5HgDk3o

This isn't a regression; As far as I can tell we've always had this behavior. 
In fact, android.util.JsonReader was changed to do it this way for consistency 
with GSON on this change: eb97c0ddc063176c26065fc6855188edf0c16e03
```

Original issue reported on code.google.com by `jessewil...@google.com` on 17 Aug 2011 at 7:40

**Comment by sshock** [comment]
Although json doesn't require escaping forward slashes, it is allowed, and it would be nice if gson had option, similar to the htmlSafe option, that would allow us to turn on escaping of forward slashes if desired.

I tried to do this myself with a custom JsonSerializer for string, but changing / to \/ doesn't work because then in the writer the \ gets escaped so we end up with \\/ instead of \/.

Now I am trying to do it with a TypeAdapter but I haven't been able to figure it out yet, sigh...

Why would I possibly want to escape forward slashes when it's not required?  The answer is I don't want to, but I have to for this special case where I am dealing with a server that I don't control that expects things a very certain way.  So yeah, not by choice...

**Comment by sshock** [comment]
This may not be the right way to do it, but here's what I came up with that seems to be working:

```java
import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;
import java.io.StringWriter;

public class EscapeSlashesAdapter extends TypeAdapter<String> {
    @Override
    public void write(JsonWriter out, String value) throws IOException {
        StringWriter sw = new StringWriter();
        JsonWriter jw = new JsonWriter(sw);
        jw.value(value);
        String escaped = sw.toString();
        escaped = escaped.replace("/", "\\/");
        out.jsonValue(escaped);
    }

    @Override
    public String read(JsonReader in) throws IOException {
        return in.nextString();
    }
}
```

**Comment by DanieleSassoli** [comment]
I've just hit this issue today with gson 2.13.2.
`gson.fromJson("test", JsonElement.class)` succeeds, but `gson.fromJson("test/test", JsonElement.class)` throws a JsonSyntaxException.

Is there no better way to fix this than registering a type adapter?

**Comment by eamonnmcmanus** [comment]
> I've just hit this issue today with gson 2.13.2. `gson.fromJson("test", JsonElement.class)` succeeds, but `gson.fromJson("test/test", JsonElement.class)` throws a JsonSyntaxException.
> 
> Is there no better way to fix this than registering a type adapter?

Can you clarify somewhat? I would not expect `gson.fromJson("test/test", JsonElement.class)` to work, since the string `test/test` is not valid JSON. (`test` is not valid JSON either. Gson allows it in lenient mode (the default) but not in strict mode.)

`gson.fromJson("\"test/test\"")` works, since the string `"test/test"` (including quotes) is valid Json.

**Comment by DanieleSassoli** [comment]
@eamonnmcmanus yeah, that was indeed the case. I guess it's a bit confusing that `"test"` works in lenient mode, but `"test/test"` doesn't. If they both didn't work I would have found it less strange.

Thanks for taking the time to reply though.

**Comment by eamonnmcmanus** [comment]
I definitely think that having lenient mode be the default was a mistake. But it's a mistake made long ago that we can't reasonably correct without risking a lot of breakage.
